File: src/audioWfc/musicTheory/NoteSet.java
package audioWfc.musicTheory;

import java.util.Objects;
import java.util.Set;

public abstract class NoteSet {
    protected Note root;
    public abstract Set<Integer> noteValues();

    public Note getRoot() {
        return this.root;
    }

    @Override
    public boolean equals(Object o){
        if(o == null) return false;
        if(this == o) return true;
        if(! (o instanceof NoteSet)) return false;
        NoteSet that = (NoteSet) o;
        return this.root.equals(that.root) &&
                this.noteValues().equals(that.noteValues());
    }

    @Override
    public int hashCode(){
        return Objects.hash(root, noteValues());
    }
}

File: src/music_theory/NoteSet.ts
import { Note, relativeNote } from "./Note"

export abstract class NoteSet {
	protected root: Note
	protected noteValues: number[]
	protected notes: Note[]

	getRoot(): Note {
		return this.root
	}

	constructor(root: Note, noteValues: number[]) {
		this.root = root
		this.noteValues = noteValues
		this.notes = noteValues.map((value) => relativeNote(root, value))
	}

	containsNote(note: Note): boolean {
		return this.notes.includes(note)
	}

	getNotes(): Note[] {
		return this.notes
	}
}
